src/snappy_device_agents/devices/maas2/maas_storage.py

```diff
@@ -0,0 +1,546 @@
+# Copyright (C) 2023 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Ubuntu MaaS 2.x storage provisioning code."""
+
+import logging
+import subprocess
+import collections
+import json
+import math
+
+
+logger = logging.getLogger()
+
+
+class MaasStorageError(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+class MaasStorage:
+    def __init__(self, maas_user, node_id, storage_data):
+        self.maas_user = maas_user
+        self.node_id = node_id
+        self.device_list = storage_data
+        self.node_info = self._node_read()
+        self.block_ids = {}
+        self.partition_sizes = {}
+
+    def _logger_debug(self, message):
+        logger.debug("MAAS: {}".format(message))
+
+    def _logger_info(self, message):
+        logger.info("MAAS: {}".format(message))
+
+    def _node_read(self):
+        """Read node block-devices.
+
+        :return: the node's block device information
+        """
+        cmd = ["maas", self.maas_user, "block-devices", "read", self.node_id]
+        return self.call_cmd(cmd, output_json=True)
+
+    @staticmethod
+    def call_cmd(cmd, output_json=False):
+        """Run a command and return the output.
+
+        :param cmd: command to run
+        :param output_json: output the result as JSON
+        :return: subprocess stdout
+        """
+        proc = subprocess.run(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, check=False
+        )
+        if proc.returncode:
+            raise MaasStorageError(proc.stdout.decode())
+
+        if proc.stdout:
+            output = proc.stdout.decode()
+
+            if output_json:
+                return json.loads(output)
+
+            return output
+
+    @staticmethod
+    def convert_size_to_bytes(size_str):
+        """Convert given sizes to bytes; case insensitive.
+
+        :param size_str: the size string to convert
+        :return: the size in bytes
+        :raises MaasStorageError: on invalid size unit/type
+        """
+        size_str = size_str.upper()
```

Comment:
You may be able to do some with 
```
power_lookup = { "T": 4, "G": 3,....,"B":, 0}
try:
    return round(float(size_str[:-1])) * (1000 ** power_lookup[size_str[-1]])
 except KeyError:
     try:
        ...
```

Reply:
That's fine. Originally looked at a similar method, but only used a couple of units to start. This way makes more sense with the expanded unit base.

Reply:
noice!